### ISSUE TITLE
mapIdentifiers() does not work with EnsDb.* packages

### ISSUE THREAD

**Issue by rcastelo** [Bioconductor/GSEABase#4]
In this [thread](https://support.bioconductor.org/p/130649) at the Bioconductor support site it came out that the function `GSEABase::mapIdentifiers()` does not play well with `EnsDb.*` annotation packages. Here is a minimal reproducible example (it requires the `EnsDb.Hsapiens.v75` package being installed:

```
download.file("https://data.broadinstitute.org/gsea-msigdb/msigdb/release/7.0/c2.cp.kegg.v7.0.symbols.gmt",  "c2.cp.kegg.v7.0.symbols.gmt")
keggsym <- getGmt("c2.cp.kegg.v7.0.symbols.gmt",  geneIdType=SymbolIdentifier())
mapIdentifiers(keggsym, ENSEMBLIdentifier("EnsDb.Hsapiens.v75"))
Error in eval(parse(text = pkg)) : 
  object 'EnsDb.Hsapiens.v75.db' not found
17: eval(parse(text = pkg))
16: eval(parse(text = pkg))
15: eval(parse(text = pkg))
14: getAnnMap(toupper(geneIdType(from)), annotation(from))
13: .mapIdentifiers_selectMaps(from, to)
12: .mapIdentifiers_map(ids, type[[1]], type[[2]], verbose)
11: mapIdentifiers(what, to, from = geneIdType(what), ..., verbose = verbose)
10: mapIdentifiers(what, to, from = geneIdType(what), ..., verbose = verbose)
9: eval(call, parent.frame())
8: eval(call, parent.frame())
7: callGeneric(what, to, from = geneIdType(what), ..., verbose = verbose)
6: FUN(X[[i]], ...)
5: FUN(X[[i]], ...)
4: lapply(what, mapIdentifiers, to, ..., verbose = verbose)
3: GeneSetCollection(lapply(what, mapIdentifiers, to, ..., verbose = verbose))
2: mapIdentifiers(keggsym, ENSEMBLIdentifier("EnsDb.Hsapiens.v75"))
1: mapIdentifiers(keggsym, ENSEMBLIdentifier("EnsDb.Hsapiens.v75"))
```

you can get around this error with the following hack, but it leads to another error:

```
EnsDb.Hsapiens.v75.db <- EnsDb.Hsapiens.v75
mapIdentifiers(keggsym, ENSEMBLIdentifier("EnsDb.Hsapiens.v75"))
Error in .select(x = x, keys = keys, columns = columns, keytype = keytype,  : 
  Argument keytype is mandatory if keys is a character vector!
21: stop("Argument keytype is mandatory if keys is a", " character vector!")
20: .select(x = x, keys = keys, columns = columns, keytype = keytype, 
        ...)
19: select(x, keys = keys(x), columns = col)
18: select(x, keys = keys(x), columns = col)
17: withCallingHandlers(expr, warning = function(w) invokeRestart("muffleWarning"))
16: suppressWarnings(tab <- select(x, keys = keys(x), columns = col))
15: AnnotationDbi:::makeFlatBimapUsingSelect(db, col = map)
14: getAnnMap(toupper(geneIdType(from)), annotation(from))
13: .mapIdentifiers_selectMaps(from, to)
12: .mapIdentifiers_map(ids, type[[1]], type[[2]], verbose)
11: mapIdentifiers(what, to, from = geneIdType(what), ..., verbose = verbose)
10: mapIdentifiers(what, to, from = geneIdType(what), ..., verbose = verbose)
9: eval(call, parent.frame())
8: eval(call, parent.frame())
7: callGeneric(what, to, from = geneIdType(what), ..., verbose = verbose)
6: FUN(X[[i]], ...)
5: FUN(X[[i]], ...)
4: lapply(what, mapIdentifiers, to, ..., verbose = verbose)
3: GeneSetCollection(lapply(what, mapIdentifiers, to, ..., verbose = verbose))
2: mapIdentifiers(keggsym, ENSEMBLIdentifier("EnsDb.Hsapiens.v75"))
1: mapIdentifiers(keggsym, ENSEMBLIdentifier("EnsDb.Hsapiens.v75"))
```

This error has been already reported in the past in Biostars in this [thread](https://www.biostars.org/p/300662), in the context of using the function `annotate::getSYMBOL()` with an `EnsDb.*` package. In both cases, the problem seems to be associated with the function `annotate::getAnnMap()`, but maybe there is a solution within `GSEABase`.

**Comment by jorainer** [comment]
Thanks @rcastelo for reporting. I will also have a look into this.

**Comment by jorainer** [comment]
The fix above (in the current 3.11 release and development branches) introduces a default for parameter `keytype` (being the Ensembl gene ID) in the `select,EnsDb` method. This fixes the error above:

```r
> library(GSEABase)
> download.file("https://data.broadinstitute.org/gsea-msigdb/msigdb/release/7.0/c2.cp.kegg.v7.0.symbols.gmt",  "c2.cp.kegg.v7.0.symbols.gmt")
> keggsym <- getGmt("c2.cp.kegg.v7.0.symbols.gmt",  geneIdType=SymbolIdentifier())
> library(EnsDb.Hsapiens.v86)
> EnsDb.Hsapiens.v86.db <- EnsDb.Hsapiens.v86
> mapIdentifiers(keggsym, ENSEMBLIdentifier("EnsDb.Hsapiens.v86"))
GeneSetCollection
  names: KEGG_GLYCOLYSIS_GLUCONEOGENESIS, KEGG_CITRATE_CYCLE_TCA_CYCLE, ..., KEGG_VIRAL_MYOCARDITIS (186 total)
  unique identifiers: ENSG00000131069, ENSG00000106633, ..., ENSG00000114867 (6524 total)
  types in collection:
    geneIdType: ENSEMBLIdentifier (1 total)
    collectionType: NullCollection (1 total)

```

**Comment by rcastelo** [comment]
Thanks Johannes for fixing this so quickly!!
